### PR TITLE
feat(events): implement manual check-in

### LINKED ISSUE #237: feat(events): participation module MVP
## Problem Statement

A He4rt Developers organiza meetups presenciais, workshops e conferências (incluindo um meetup em breve e uma conferência de 2 dias no final do ano), mas não possui um sistema próprio para gerenciar o ciclo de participação dos membros. Hoje não há como:
- Controlar inscrições com capacidade limitada e lista de espera
- Verificar presença de quem realmente compareceu (check-in)
- Recompensar participação com XP no sistema de gamificação existente
- Aprovar candidatos a eventos com vagas limitadas (workshops)

## Solution

Construir o módulo de participação em eventos (`app-modules/events/`) com lifecycle completo: enrollment → check-in → attendance → XP reward. O sistema suporta 3 métodos de entrada (RSVP, RSVP+check-in, Application com formulário dinâmico), 3 métodos de check-in (manual, código numérico, QR code), waitlist com promoção FIFO, e integração com o módulo de gamificação via domain events.

## User Stories

1. As a participant, I want to RSVP to a meetup with one click, so that I can confirm my presence quickly
2. As a participant, I want to join a waitlist when an event is full, so that I can be promoted if someone cancels
3. As a participant, I want to be automatically promoted from the waitlist when a slot opens, so that I don't miss the opportunity
4. As a participant, I want to cancel my enrollment before the deadline, so that my slot is freed for others
5. As a participant, I want to check in at a meetup by typing a numeric code, so that my presence is verified
6. As a participant, I want to check in at a conference by showing my QR code badge, so that attendance is tracked per day
7. As a participant, I want to receive XP when I confirm, check in, and complete attendance, so that my participation is rewarded in the gamification system
8. As a participant, I want to fill an application form to apply for a workshop, so that the organizer can evaluate my candidacy
9. As a participant, I want to be notified when my application is approved or rejected, so that I know my status
10. As a participant, I want to check in on each day of a multi-day conference, so that my attendance per day is tracked
11. As a participant, I want to use the same QR code across all event days, so that I only need one badge
12. As a participant, I want to check in via Discord/Twitch bot by typing a command with the code, so that I can confirm presence in online events
13. As a participant, I want to see my enrollment status for each event, so that I know where I stand
14. As an organizer, I want to create events with configurable enrollment policies, so that each event has appropriate participation rules
15. As an organizer, I want to set capacity limits with optional waitlist, so that I can control attendance
16. As an organizer, I want to choose the enrollment method (RSVP, RSVP+check-in, Application), so that the entry flow matches the event type
17. As an organizer, I want to choose the check-in method (manual, numeric code, QR code), so that presence verification fits the event format
18. As an organizer, I want to define a dynamic application form (JSONB schema), so that I can collect relevant information per event
19. As an organizer, I want to approve or reject applications with a reason, so that participants understand the decision
20. As an organizer, I want to generate time-limited numeric codes bound to a specific date, so that check-in is controlled per day
21. As an organizer, I want to manually check in participants, so that I can handle edge cases (late arrivals, code issues)
22. As an organizer, I want to override a participant's status (e.g., no_show → attended), so that I can correct errors
23. As an organizer, I want to set a cancellation deadline in hours before the event, so that I have predictability on attendance
24. As an organizer, I want to configure XP rewards per event (RSVP, check-in, attendance), so that different events reward differently
25. As an organizer, I want to set attendance_requirement (all_days, any_day, minimum_days), so that multi-day events have clear completion criteria
26. As an organizer, I want to see all enrollments with their current status, so that I can manage the event
27. As an organizer, I want a full audit trail of every status transition, so that I can investigate disputes
28. As the system, I want to automatically mark checked_in → attended after the event ends, so that XP is awarded without manual intervention
29. As the system, I want to automatically mark confirmed → no_show after the event ends, so that non-attendees are identified
30. As the system, I want enrollment capacity checks to be atomic (lockForUpdate), so that overbooking is impossible under concurrency

### Architecture

- **Module**: `app-modules/events/` following existing modular architecture (`internachi/modular`)
- **Namespace**: `He4rt\Events\*`
- **Panels**: Admin panel = organizer, App panel = participant
- **Table prefix**: All tables prefixed with `events_` for clear identification in database

### State Machine

- Implemented as a PHP backed enum (`EnrollmentStatusEnum`) with `canTransitionTo()` method
- **NOT** using `spatie/laravel-model-states` — transitions validated in Action classes explicitly
- 8 states: `pending`, `confirmed`, `waitlisted`, `checked_in`, `attended`, `cancelled`, `rejected`, `no_show`
- Terminal states: `attended` (success), `cancelled`, `rejected`, `no_show`
- See ADR-0001

### Schema (7 tables)

**`events`** — Core event record:
- `id`, `tenant_id`, `event_type` (meetup/workshop/conference), `active`, `slug`, `title`, `description`, `location`, `starts_at`, `ends_at`, `timestamps`
- No denormalized counters — derived by query

**`events_enrollment_policies`** — 1:1 configuration per event:
- `event_id` (unique), `enrollment_method`, `check_in_method`, `capacity` (nullable = unlimited), `has_waitlist`, `attendance_requirement` (all_days/any_day/minimum_days), `cancellation_deadline_hours`, `xp_reward_rsvp`, `xp_reward_checkin`, `xp_reward_attendance`, `application_form_schema` (JSONB), `timestamps`

**`events_enrollments`** — Hub central (1 per user+event):
- `id`, `event_id`, `user_id`, `status` (enum), `waitlist_position` (nullable), `application_data` (JSONB), `rejection_reason`, `enrolled_at`, `confirmed_at`, `checked_in_at`, `attended_at`, `cancelled_at`, `is_public`, `timestamps`
- UNIQUE constraint on `(event_id, user_id)`

**`events_enrollment_transitions`** — Append-only audit trail:
- `id`, `enrollment_id`, `from_status`, `to_status`, `triggered_by` (system/user/admin), `actor_user_id` (nullable), `reason` (nullable), `created_at`
- Written explicitly by Actions, NOT by database triggers (ADR-0003)

**`events_check_ins`** — 1:N per enrollment (one per day):
- `id`, `enrollment_id`, `method` (manual/numeric_code/qr_code), `payload` (JSONB), `event_date`, `checked_in_at`, `timestamps`
- Multi-day events: one check-in record per day attended (ADR-0004)
- Event days derived from `events.starts_at`/`ends_at` date range — no separate days table

**`events_check_in_codes`** — Organizer-generated numeric codes:
- `id`, `event_id`, `code`, `event_date`, `valid_from`, `valid_until`, `max_uses` (nullable), `uses_count`, `timestamps`
- Bound to specific date — one code per day

**`events_qr_tokens`** — One per enrollment, reusable across days:
- `id`, `enrollment_id`, `token` (unique), `expires_at` (nullable), `timestamps`

### XP Integration

- Events module publishes domain events: `EnrollmentConfirmed`, `ParticipantCheckedIn`, `ParticipantAttended`
- Gamification module listens and calls `IncrementExperience` on Character
- XP amounts come from `events_enrollment_policies` fields, passed as payload in domain events
- No XP ledger in Events module — auditability via `events_enrollment_transitions`
- Idempotency responsibility on Gamification listener
- See ADR-0002

### Bot Integration

- Discord/Twitch bot dispatches a domain event (e.g., `CheckInRequested`) with user identifier and code
- Events module has a listener that validates the code and processes the check-in
- Bot is pure transport — all business logic in Events module
- Bot must listen for response event (`CheckInProcessed`) to reply to user in chat
- Consistent with existing Bot Discord → Moderation pattern in CONTEXT-MAP
- See ADR-0005

### Enrollment Flow — Concurrency

- Capacity check uses `DB::transaction` with `lockForUpdate()` on the enrollment policy
- Count confirmed/checked_in/attended enrollments → compare against capacity → insert or waitlist
- Waitlist promotion on cancellation: FIFO (ordered by `enrolled_at`), atomic update + notification dispatch

### Scheduled Job

- Runs after `event.ends_at` passes
- Batch transitions: `checked_in → attended` (if attendance_requirement met), `confirmed → no_show`
- Dispatches XP domain events for newly `attended` enrollments
- Single job handles both transitions per event

### Application Form

- Dynamic form schema stored as JSONB in `events_enrollment_policies.application_form_schema`
- Filament form builder for organizer to define questions
- Participant responses stored in `events_enrollments.application_data` as JSONB
- Organizer reviews in Admin panel with approve/reject actions

### Event Types

```php
enum EventTypeEnum: string {
    case Meetup = 'meetup';
    case Workshop = 'workshop';
    case Conference = 'conference';
}
```

## Testing Decisions

- **Good tests**: test external behavior through Actions and domain events, not internal implementation. Assert state transitions, database state, and dispatched events.
- **Test location**: `app-modules/events/tests/Feature/` and `app-modules/events/tests/Unit/`
- **Prior art**: existing tests in `app-modules/moderation/tests/` and `app-modules/bot-discord/tests/`
- **What to test**:
  - Each Action (EnrollUser, CancelEnrollment, CheckIn, ApproveApplication, RejectApplication, PromoteFromWaitlist) — happy path + edge cases
  - State machine transitions — valid and invalid
  - Concurrency: capacity enforcement under race conditions
  - Waitlist promotion FIFO ordering
  - Multi-day check-in with attendance_requirement validation
  - Scheduled job (attended + no_show marking)
  - Domain event dispatch (assert events are fired, don't test Gamification's listener here)
  - Cancellation deadline enforcement
  - Numeric code validation (expiry, max_uses, wrong date)
  - QR token validation (expiry, reuse across days)
- **Factories**: create factories for EventModel, EventEnrollmentPolicy, EventEnrollment, EventCheckIn, EventCheckInCode, EventQrToken
- **Use Pest** with `php artisan make:test --pest`
- **ActingAs**: all Filament panel tests require `$this->actingAs(User::factory()->create())`

## Out of Scope

- Networking between participants (profiles, connections, discovery)
- Referral / invite links with tracking
- Magic link check-in method
- Geolocation check-in method
- Sponsors association with events
- Timeline / activity feed display
- Call for Papers / Submissions (CFP)
- Agenda / schedule display
- Paid events / payment integration
- No-show penalties affecting future eligibility (data tracked, consequences deferred)
- Event days as explicit table (derived from date range)
- Denormalized counters on events table

## Further Notes

- The events module was previously gutted (all tables dropped in migration `2026_05_16_195205`). This PRD rebuilds the module from scratch with the new participation-focused architecture.
- The previous module had: events, events_talks, events_attendees, sponsors, events_sponsors, events_agenda, event_submission_speakers. The new module replaces this with a fundamentally different schema focused on enrollment lifecycle.
- Existing ADRs document key architectural decisions: see `app-modules/events/docs/adr/0001-0005`.
- `CONTEXT-MAP.md` at repo root has been updated to include Events and Gamification bounded contexts with dependency diagram.
- `app-modules/events/CONTEXT.md` contains the full domain glossary and state machine reference.
- Future phases (post-MVP): magic_link/geolocation check-in, CFP/submissions, agenda display, networking, referral, payment integration, no-show penalties.

### LINKED ISSUE #244: feat(events): manual check-in
## Parent

#237

## What to build

Implement manual check-in: organizer marks a participant as present via Admin panel. Creates a check-in record and transitions enrollment status.

**Action: `CheckInAction`**
- Accepts: enrollment, method (enum), payload (array), event_date (Carbon date)
- Validates: enrollment status must be `confirmed`
- Validates: event_date is within event's starts_at/ends_at date range
- Validates: no existing check-in for this enrollment + date combination
- Creates `events_check_ins` record (method=manual, payload={actor_user_id: ...}, event_date, checked_in_at=now)
- If this is the first check-in for this enrollment: transitions status from `confirmed` → `checked_in`, sets `checked_in_at`
- If enrollment is already `checked_in` (multi-day, subsequent days): just creates the check-in record, no status transition
- Records transition in audit trail (triggered_by=admin, actor_user_id=organizer)
- Dispatches `ParticipantCheckedIn` implementing `ShouldDispatchAfterCommit`

**Admin panel:**
- "Check In" action on enrollment records (only visible when status=confirmed or checked_in)
- Date picker defaulting to today (for multi-day events)
- Bulk action: "Check In Selected" for marking multiple participants at once
- Check-in history visible per enrollment (dates checked in)

**Validation rules:**
- Cannot check in the same enrollment on the same date twice
- Cannot check in outside event date range
- Cannot check in enrollment that is not `confirmed` or `checked_in`

## Acceptance criteria

- [ ] Organizer can mark a participant as checked in from Admin panel
- [ ] Check-in record is created in `events_check_ins` with method=manual and correct date
- [ ] First check-in transitions enrollment from `confirmed` → `checked_in`
- [ ] Subsequent check-ins (multi-day) create records without re-transitioning status
- [ ] Duplicate check-in on same date is rejected
- [ ] Check-in outside event date range is rejected
- [ ] Bulk check-in action works for multiple participants
- [ ] `ParticipantCheckedIn` domain event is dispatched
- [ ] Transition audit trail records the organizer who performed check-in
- [ ] Feature tests: single check-in, multi-day check-ins, duplicate prevention, date range validation
- [ ] Pint passes

## Blocked by

- #240 (RSVP Enrollment — needs confirmed enrollments to check in)

### PR DESCRIPTION
**Pull request by davicbtoliveira**
### Solve: #244 

Implements manual check-in: organizers mark participants as present via the Admin panel. Creates a check-in record and transitions enrollment status. Covers single-day and multi-day events with duplicate prevention, date range validation, audit trail, and bulk check-in.

### `CheckInAction` — Core action

- Accepts enrollment, method (enum), payload (array), event_date (Carbon date)
- Validates enrollment status is `confirmed` or `checked_in`
- Validates event_date is within the event's date range
- Validates no existing check-in for the same enrollment + date
- Creates `events_check_ins` record with `method=manual`, `payload={actor_user_id}`, `event_date`, `checked_in_at=now`
- First check-in transitions enrollment from `confirmed` → `checked_in`
- Subsequent check-ins (multi-day) only create the check-in record
- Records audit trail (`triggered_by=admin`, `actor_user_id=organizer`)
- Dispatches `ParticipantCheckedIn` domain event (`ShouldDispatchAfterCommit`)

### Admin panel — `EnrollmentsRelationManager`

- "Check In" action on enrollment records (visible when status is `confirmed` or `checked_in`)
- Date picker defaulting to today, constrained to event date range
- Bulk "Check In Selected" action for multiple participants
- Check-in history column per enrollment

### `TransitionEnrollmentAction`

Extracted reusable action for enrollment status transitions, used by the check-in flow and available for future transitions (`attended`, `cancelled`).

### `CheckInException`

Check-in validation errors moved to the CheckIn module boundary (`CheckIn/Exceptions/`) instead of leaking into `EnrollmentException`.

## Acceptance criteria

- [x] Organizer can mark a participant as checked in from Admin panel
- [x] Check-in record created in `events_check_ins` with method=manual and correct date
- [x] First check-in transitions enrollment from `confirmed` → `checked_in`
- [x] Subsequent check-ins (multi-day) create records without re-transitioning
- [x] Duplicate check-in on same date is rejected
- [x] Check-in outside event date range is rejected
- [x] Bulk check-in action works for multiple participants
- [x] `ParticipantCheckedIn` domain event is dispatched
- [x] Transition audit trail records the organizer
- [x] Feature tests: single, multi-day, duplicate, date range, status validation
- [x] Pint passes

### New (8 files)

| File | Purpose |
|---|---|
| `app-modules/events/src/CheckIn/Actions/CheckInAction.php` | Core check-in action (validate, create record, transition, dispatch) |
| `app-modules/events/src/CheckIn/Exceptions/CheckInException.php` | Check-in specific exceptions (invalid status, outside range, duplicate, invalid actor) |
| `app-modules/events/src/CheckIn/Events/ParticipantCheckedIn.php` | Domain event dispatched after check-in (`ShouldDispatchAfterCommit`) |
| `app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php` | Reusable action for enrollment status transitions with audit trail |
| `app-modules/events/lang/en/check_in.php` | English check-in exception messages |
| `app-modules/events/lang/pt_BR/check_in.php` | Brazilian Portuguese check-in exception messages |
| `app-modules/events/database/factories/CheckInFactory.php` | Factory for CheckIn model |
| `app-modules/events/tests/Feature/CheckInActionTest.php` | 6 feature tests (single, multi-day, duplicate, date range, status, actor) |

### Modified (11 files)

| File | What changed |
|---|---|
| `app-modules/events/src/CheckIn/Models/CheckIn.php` | Added `checked_in_at` to fillable, casts, phpdocs |
| `app-modules/events/database/migrations/2026_05_22_000001_add_checked_in_at_to_events_check_ins_table.php` | Added `checked_in_at` column to `events_check_ins` |
| `app-modules/events/src/Enrollment/Models/Enrollment.php` | Added `checkIns()` hasMany relationship |
| `app-modules/events/src/Enrollment/Enums/EnrollmentStatus.php` | Added `CheckedIn` case with transition rules |
| `app-modules/events/src/Enrollment/Exceptions/EnrollmentException.php` | Added `invalidTransition()`; removed check-in exceptions (moved to CheckInException) |
| `app-modules/events/src/Enrollment/Actions/EnrollUserAction.php` | Import adjustment |
| `app-modules/events/src/Event/Models/Event.php` | Relationship additions for enrollment policy |
| `app-modules/events/lang/en/exceptions.php` | Added `invalid_transition` key; removed check-in keys |
| `app-modules/events/lang/pt_BR/exceptions.php` | Added `invalid_transition` key; removed check-in keys |
| `app-modules/events/tests/Feature/EventResourceTest.php` | 5 tests: Filament CRUD, single + bulk check-in, check-in history |
| `app-modules/panel-admin/src/Filament/Resources/Events/RelationManagers/EnrollmentsRelationManager.php` | "Check In" action, bulk "Check In Selected", check-in history column, date picker |

### CODE REVIEW

File: app-modules/events/src/CheckIn/Actions/CheckInService.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+use He4rt\Events\CheckIn\Models\CheckIn;
+use He4rt\Events\Enrollment\Actions\TransitionEnrollmentAction;
+use He4rt\Events\Enrollment\Enums\EnrollmentStatus;
+use He4rt\Events\Enrollment\Enums\TriggeredBy;
+use He4rt\Events\Enrollment\Models\Enrollment;
+use Illuminate\Database\UniqueConstraintViolationException;
+use Illuminate\Support\Facades\Date;
+use Illuminate\Support\Facades\DB;
+
+final readonly class CheckInService
```
Comment (danielhe4rt) [inline app-modules/events/src/CheckIn/Actions/CheckInService.php]:
```suggestion
final readonly class CheckInAction
```

File: app-modules/events/src/CheckIn/Actions/CheckInService.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    /**
+     * @param  array<string, mixed>  $payload
+     */
+    public function handle(
+        Enrollment $enrollment,
+        CheckInMethod $method,
+        array $payload,
+        CarbonInterface $eventDate,
+        string $actorUserId,
+        TriggeredBy $triggeredBy,
```
Comment (danielhe4rt) [inline app-modules/events/src/CheckIn/Actions/CheckInService.php]: receives a DTO with all needed information and constraints.

goal: wrap it into a CheckInDTO

File: app-modules/events/src/CheckIn/Actions/ManualCheckInAction.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+use He4rt\Events\CheckIn\Models\CheckIn;
+use He4rt\Events\Enrollment\Enums\TriggeredBy;
+use He4rt\Events\Enrollment\Models\Enrollment;
+
+final readonly class ManualCheckInAction
+{
+    public function handle(
+        Enrollment $enrollment,
+        string $actorUserId,
+        CarbonInterface $eventDate,
```
Comment (danielhe4rt) [inline app-modules/events/src/CheckIn/Actions/ManualCheckInAction.php]: ditto: dto

File: app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+     * @param  array<string, mixed>  $metadata
+     */
+    public function handle(
+        Enrollment $enrollment,
+        EnrollmentStatus $toStatus,
+        TriggeredBy $triggeredBy,
+        ?string $actorId = null,
+        ?string $reason = null,
+        array $metadata = [],
+        ?CarbonInterface $timestamp = null,
```
Comment (danielhe4rt) [inline app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php]: dto

File: app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            'triggered_by' => $triggeredBy,
+            'reason' => $reason,
+            'metadata' => $metadata ?: null,
+        ]);
+    }
+
+    private function resolveTimestampColumn(EnrollmentStatus $status): ?string
+    {
+        return self::TIMESTAMP_MAP[$status->value] ?? null;
+    }
```
Comment (danielhe4rt) [inline app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php]: move this to the dto.

File: app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+use He4rt\Events\Enrollment\Models\EnrollmentTransition;
+
+final readonly class TransitionEnrollmentAction
+{
+    private const array TIMESTAMP_MAP = [
+        'confirmed' => 'confirmed_at',
+        'checked_in' => 'checked_in_at',
+        'attended' => 'attended_at',
+        'cancelled' => 'cancelled_at',
+    ];
```
Comment (danielhe4rt) [inline app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php]: this as well.

actually it wouldn't be even needed inside of it. why this mapping exists?

Comment (danielhe4rt) [inline app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php]: remove it and move to the enum or the dto. pick your poison

**Review by danielhe4rt — CHANGES_REQUESTED** [review]
tá maneiro doggers, só arruma umas coisas ai

File: app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            $fromStatus->canTransitionTo($toStatus),
+            EnrollmentException::invalidTransition($fromStatus, $toStatus),
+        );
+
+        $timestampColumn = $toStatus->timestampColumn();
+
+        $dto->enrollment->forceFill([
+            'status' => $toStatus,
+            $timestampColumn => $dto->timestamp ?? now(),
+        ])->save();
```
Comment (danielhe4rt) [inline app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php]:
```suggestion
        $dto->enrollment->create([
            'status' => $toStatus,
            $timestampColumn => now(),
        ]);
```

File: app-modules/events/src/Enrollment/Enums/EnrollmentStatus.php [hunk 8]
```diff
@@ -90,6 +98,7 @@ public function getResponseMessage(): string
         return match ($this) {
             self::Confirmed => __('events::pages.confirm_presence_success'),
             self::Waitlisted => __('events::pages.waitlist_success'),
+            default => '',
```
Comment (danielhe4rt) [inline app-modules/events/src/Enrollment/Enums/EnrollmentStatus.php]: Geralmente quando cai num caso que não é listado, ele deveria retornar uma Exception com algo como "notImplemented()". Voltar uma string vazia é pedir pra se fuder num possível debug. (leia com a voz do seu primo)

File: app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php (hunk 7 above)
Comment (davicbtoliveira) [reply app-modules/events/src/Enrollment/Actions/TransitionEnrollmentAction.php]: Done. I kept audit correctness as the source of truth:
  every status change is timestamped in
  `events_enrollment_transitions`.

  The `*_at` fields on `events_enrollments` stay as
  convenience milestone columns only, so the action now
  updates them conditionally when the target status has one.

File: app-modules/events/src/Enrollment/Enums/EnrollmentStatus.php (hunk 8 above)
Comment (davicbtoliveira) [reply app-modules/events/src/Enrollment/Enums/EnrollmentStatus.php]: Done by creating a new Exception.

**Review by danielhe4rt — APPROVED** [review]
